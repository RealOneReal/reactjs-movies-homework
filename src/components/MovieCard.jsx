import {
  Card,
  Typography,
  CardMedia,
  CardContent,
  Avatar,
  Box,
} from "@mui/material";
import { PlayCircle } from "@mui/icons-material";
import { truncateString } from "../utils/stringUtils";

const MovieCard = (props) => {
  const { poster_path, genre_ids, vote_average, original_title, id } = props;
  const title = truncateString(original_title, 19);
  const genres = truncateString(genre_ids?.length && genre_ids.join(" "), 26);

  const handleClick = () => {
    console.log(id);
  };

  const styles = {
    card: {
      width: "200px",
      height: "300px",
      position: "relative",
      bgcolor: "secondary.dark",
      ":hover": {
        transition: "0.6s ease",
        opacity: 0.8,
      },
      ":hover .rating": {
        display: "none",
      },
    },
    hoverLayer: {
      width: "inherit",
      height: "220px",
      position: "absolute",
      bgcolor: "secondary.dark",
      opacity: 0,
      textAlign: "center",
      pt: "60%",
      cursor: "pointer",
      ":hover": {
        transition: " .3s ease",
        opacity: 1,
      },
    },
    playIcon: { color: "primary.light", fontSize: "60px" },
    ratingBox: {
      position: "absolute",
      top: "20px",
      left: "0px",
      borderRadius: "5px",
      height: "25px",
      width: "40px",
      bgcolor: "green",
      textAlign: "center",
    },
    cardImage: {
      width: "180px",
      ml: 1,
      mt: 1,
      borderRadius: "10px",
    },
    genres: { display: "flex", textAlign: "center", columnGap: 1 },
  };

  return (
    <Card sx={styles.card}>
      <Box sx={styles.hoverLayer} onClick={handleClick}>
        <PlayCircle sx={styles.playIcon} />
      </Box>

      <Box className="rating" sx={styles.ratingBox}>
        <Typography variant="subtitle1" component="span">
          {vote_average}
        </Typography>
      </Box>
      <CardMedia
        component="img"
        height="220px"
        image={poster_path}
        sx={styles.cardImage}
      />
      <CardContent sx={{ p: 1 }}>
        <Typography variant="h6" color="primary.light">
          {title}
        </Typography>
        <Typography variant="subtitle2" sx={styles.genres} color="primary.dark">
          {genres}
        </Typography>
      </CardContent>
    </Card>
  );
};

export default MovieCard;
