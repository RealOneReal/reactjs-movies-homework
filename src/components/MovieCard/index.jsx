import { Card, Typography, CardMedia, CardContent, Box } from "@mui/material";
import { PlayCircle } from "@mui/icons-material";
import { truncateString } from "../../utils/stringUtils";

const MovieCard = (props) => {
  const { poster_path, genre_ids, vote_average, original_title, id } = props;
  const title = truncateString(original_title, 19);
  const genres = truncateString(genre_ids?.length && genre_ids.join(" "), 26);

  const handleClick = () => {
    console.log(id);
  };

  const styles = {
    container: {
      position: "relative",
      ":hover .MuiPaper-root": {
        transition: "0.6s ease",
        opacity: 0.3,
      },
    },
    card: {
      width: "200px",
      height: "300px",
      position: "relative",
      bgcolor: "secondary.dark",
    },
    hoverLayer: {
      width: "200px",
      height: "300px",
      position: "absolute",
      left: 0,
      top: 0,
      opacity: 0,
      textAlign: "center",
      cursor: "pointer",
      ":hover": {
        transition: " .3s ease",
        opacity: 1,
      },
    },
    playIcon: {
      color: "primary.main",
      fontSize: "60px",
      position: "absolute",
      top: "calc(50% - 30px)",
      left: "calc(50% - 30px)",
    },
    ratingBox: {
      position: "absolute",
      top: "20px",
      left: "0px",
      borderRadius: "5px",
      height: "25px",
      width: "40px",
      bgcolor: "green",
      textAlign: "center",
    },
    cardImage: {
      width: "180px",
      ml: 1,
      mt: 1,
      borderRadius: "10px",
    },
    genres: { display: "flex", textAlign: "center", columnGap: 1 },
  };

  return (
    <Box sx={styles.container}>
      <Card sx={styles.card}>
        <Box className="rating" sx={styles.ratingBox}>
          <Typography variant="subtitle1" component="span">
            {vote_average}
          </Typography>
        </Box>
        <CardMedia
          component="img"
          height="220px"
          image={poster_path}
          sx={styles.cardImage}
        />
        <CardContent sx={{ p: 1 }}>
          <Typography variant="h6" color="primary.light">
            {title}
          </Typography>
          <Typography
            variant="subtitle2"
            sx={styles.genres}
            color="primary.dark"
          >
            {genres}
          </Typography>
        </CardContent>
      </Card>
      <Box sx={styles.hoverLayer} onClick={handleClick}>
        <PlayCircle sx={styles.playIcon} />
      </Box>
    </Box>
  );
};

export default MovieCard;
