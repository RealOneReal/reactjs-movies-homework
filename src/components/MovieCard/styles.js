const styles = {
  container: {
    position: "relative",
    ":hover .MuiPaper-root": {
      transition: "0.6s ease",
      opacity: 0.3,
    },
  },
  card: {
    width: "200px",
    height: "300px",
    position: "relative",
    bgcolor: "secondary.dark",
  },
  hoverLayer: {
    width: "200px",
    height: "300px",
    position: "absolute",
    left: 0,
    top: 0,
    opacity: 0,
    textAlign: "center",
    cursor: "pointer",
    ":hover": {
      transition: " .3s ease",
      opacity: 1,
    },
  },
  playIcon: {
    color: "primary.main",
    fontSize: "60px",
    position: "absolute",
    top: "calc(50% - 30px)",
    left: "calc(50% - 30px)",
  },
  ratingBox: {
    position: "absolute",
    top: "20px",
    left: "0px",
    borderRadius: "5px",
    height: "25px",
    width: "40px",
    bgcolor: "green",
    textAlign: "center",
  },
  cardImage: {
    width: "180px",
    height: "220px",
    ml: 1,
    mt: 1,
    borderRadius: "10px",
  },
  genres: { display: "flex", textAlign: "center", columnGap: 1 },
};

export default styles;
