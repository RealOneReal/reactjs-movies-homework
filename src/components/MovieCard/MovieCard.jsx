import { Card, Typography, CardMedia, CardContent, Box } from "@mui/material";
import { PlayCircle } from "@mui/icons-material";
import fallback from "../../assets/broken-image.png";
import styles from "./styles";
import { truncateString } from "../../utils/stringUtils";

const MovieCard = ({ posterPath, genreIds, voteAverage, title, id }) => {
  const movieTitle = truncateString(title, 19);
  const genres = truncateString(genreIds.join(" "), 26);

  const handleClick = () => {
    console.log(id);
  };

  const handleMediaFallback = ({ target }) => {
    target.style.objectFit = "contain";
    return (target.src = fallback);
  };

  return (
    <Box sx={styles.container} onClick={handleClick}>
      <Card sx={styles.card}>
        <Box className="rating" sx={styles.ratingBox}>
          <Typography variant="subtitle1" component="span">
            {voteAverage}
          </Typography>
        </Box>
        <CardMedia
          component="img"
          alt={movieTitle}
          src={posterPath}
          sx={styles.cardImage}
          onError={handleMediaFallback}
        />
        <CardContent sx={{ p: 1 }}>
          <Typography variant="h6" color="primary.light">
            {movieTitle}
          </Typography>
          <Typography
            variant="subtitle2"
            sx={styles.genres}
            color="primary.dark"
          >
            {genres}
          </Typography>
        </CardContent>
      </Card>
      <Box sx={styles.hoverLayer}>
        <PlayCircle sx={styles.playIcon} />
      </Box>
    </Box>
  );
};

export default MovieCard;
