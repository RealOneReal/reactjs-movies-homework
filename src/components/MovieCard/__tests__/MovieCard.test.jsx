import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import MovieCard from "../MovieCard";
import userEvent from "@testing-library/user-event";

const mockData = {
  original_title: "Ninja",
  genre_ids: ["comedy", "history"],
  id: 123,
  vote_average: 9,
};
jest.spyOn(window.console, "log");

describe("MovieCard", () => {
  it("should be render", () => {
    const tree = renderer.create(<MovieCard {...mockData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should hover and be clickable", () => {
    render(<MovieCard {...mockData} />);
    const card = screen.getByText(/Ninja/i);
    userEvent.click(card);
    expect(console.log).toBeCalled();
  });
});
