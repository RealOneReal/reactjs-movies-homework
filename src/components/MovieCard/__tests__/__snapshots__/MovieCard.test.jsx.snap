// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should be render 1`] = `
<div
  className="MuiBox-root css-1fxpa70"
  onClick={[Function]}
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1tm8ee0-MuiPaper-root-MuiCard-root"
  >
    <div
      className="rating MuiBox-root css-10gsq3s"
    >
      <span
        className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
      >
        9
      </span>
    </div>
    <img
      alt="Ninja"
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1qopqf2-MuiCardMedia-root"
      height="220px"
      onError={[Function]}
    />
    <div
      className="MuiCardContent-root css-71xg00-MuiCardContent-root"
    >
      <h6
        className="MuiTypography-root MuiTypography-h6 css-100kqv6-MuiTypography-root"
      >
        Ninja
      </h6>
      <h6
        className="MuiTypography-root MuiTypography-subtitle2 css-116y7rw-MuiTypography-root"
      >
        comedy history
      </h6>
    </div>
  </div>
  <div
    className="MuiBox-root css-lx8xrd"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-189a69h-MuiSvgIcon-root"
      data-testid="PlayCircleIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM9.5 16.5v-9l7 4.5-7 4.5z"
      />
    </svg>
  </div>
</div>
`;
